mailscanner/functions.php

```diff
@@ -3895,73 +3916,93 @@ function ($val) {
-        $graph->SetShadow();
-        $graph->img->SetAntiAliasing();
-        $graph->title->SetFont(FF_DV_SANSSERIF, FS_BOLD, 14);
-        $graph->title->Set($reportTitle);
-
-        $plotData = $data[$graphColumn['dataColumn']];
-        $legendData = $data[$graphColumn['labelColumn']];
-        $p1 = new PiePlot3d($plotData);
-        $p1->SetTheme('sand');
-        $p1->SetLegends($legendData);
-
-        $p1->SetCenter(0.7, 0.5);
-        $graph->legend->SetLayout(LEGEND_VERT);
-        $graph->legend->Pos(0.0, 0.25, 'left');
-
-        $graph->Add($p1);
-        $graph->Stroke($filename);
-
-        //  Check Permissions to see if the file has been written and that apache to read it.
-        if (is_readable($filename)) {
-            echo '<tr><td style="text-align: center"><IMG SRC="' . $filename . '" alt="Graph"></td></tr>';
-        } else {
-            echo '<tr><td style="text-align: center">' . __('message199') . ' ' . CACHE_DIR . ' ' . __('message299') . '</td></tr>';
-        }
-    } else {
-        echo '<tr><td class="center">' . sprintf(__('errorcachedirnotwritable03'), CACHE_DIR) . '</td></tr>';
-    }
-
-    echo '<tr>';
+    //create canvas graph
+    $bgcolors = getHexColors(count($data[$graphColumn['dataColumn']]));
+    echo '<canvas id="reportChart" class="reportGraph"></canvas>
+  <script src="js/Chart.js/Chart.min.js"></script>
+  <script>
+    var ctx = document.getElementById("reportChart");
+    var myChart = new Chart(ctx, {
+      type: "pie",
+      data: {
+        labels: ["' . implode('", "', $data[$graphColumn['labelColumn']]) . '"],
+        datasets: [{
+          label: "' . $reportTitle . '",
+          data: [' . implode(', ', $data[$graphColumn['dataColumn']]) . '],
+          backgroundColor: ["' . implode('", "', $bgcolors) . '"]
+        }]
+      },
+      options: {
+        title: {
+          display: true,
+          text: "' . $reportTitle . '"
+        },
+        legend: {
+          display: true,
+          position: "top"
+        },
+        responsive: false
+      }
+    });
+  </script>';
 
     // HTML to display the table
     echo '<table class="reportTable">';
     echo '    <tr style="background-color: #F7CE4A">' . "\n";
-    foreach ($columnTitles as $columnTitle) {
+    foreach ($columns as $columnName => $columnTitle) {
         echo '     <th>' . $columnTitle . '</th>' . "\n";
     }
     echo '    </tr>' . "\n";
 
     for ($i = 0; $i < $numResult; $i++) {
         echo '    <tr style="background-color: #EBEBEB">' . "\n";
-        foreach ($sqlColumns as $sqlColumn) {
-            if (isset($valueConversions[$sqlColumn])) {
-                echo '     <td>' . $data[$sqlColumn . 'conv'][$i] . '</td>' . "\n";
-            } else {
-                echo '     <td>' . $data[$sqlColumn][$i] . '</td>' . "\n";
-            }
+        foreach ($columns as $columnName => $columnTitle) {
+            echo '     <td>' . $data[$columnName][$i] . '</td>' . "\n";
```

Comment:
here is missing formatting of numbers and conversion of size from bytes to human readable units